---
title: "BuscoToCelegans"
author: "Kourosh"
date: "November 8, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Busco to C.elegans genes

What we found out was that busco uses a hmmer consensus sequence which makes it very hard for reads to align. Thus we must get the closest ancestor being C. elegans to make this work. So I am going to be taking the othrodb groups and extracting the c.elegans genes from it. Hopefully it has all 982 genes. C.elegans species number is 6239. 

```{r,eval=FALSE,include=FALSE}
library(seqinr)

orthoDbToOrganism <- read.table("ogs.id.info",stringsAsFactors = FALSE)
celegans <- orthoDbToOrganism[grep("6239:*",orthoDbToOrganism$V1),]

length(unique(celegans$V2))
#As the length mathces the number of protiens it is safe to assume that the protiens are all there. So now to extra the genes.

OrganismToSequence <-  read.fasta("6239.fs",seqtype = "AA")

sequence <- OrganismToSequence[celegans$V1]
names(sequence) <- celegans$V2


write.fasta(sequence,names(sequence),file.out = "BuscoCelegans.fasta")


```


## Analyzing the shift


```{r}
library(GenomicRanges)
library(data.table)
library(plyr)
library(dplyr)
library(seqinr)
calculateTotalCoverage <- function(coverage){
  totalCoverage <- coverage[,3]
  totalCoverage <- totalCoverage[(totalCoverage > 10 & totalCoverage<=200)]   #Limited the data so it is easier to see the main parts of it. We can also limit the data later on as well. 
  totalCoverage <- as.data.frame(totalCoverage)
  return(totalCoverage)
}

getmode <- function(v) {
  uniqv <- unique(v)
  uniqv[which.max(tabulate(match(v, uniqv)))]
}
library(ggplot2)


PercentAAGreaterThan <- function(diamondGRanges,proteinFile){
  coverage <- as.data.frame(coverage(diamondGRanges))
  totalCoverage <- calculateTotalCoverage(coverage)
  threholdCoverage <- getmode(totalCoverage[,1])
  
  pertcentAAGreaterThanThreshold <-   coverage[,c(2,3)] %>%
      group_by(group_name) %>%
      filter(value >= threholdCoverage)%>%
      tally()
    pertcentAAGreaterThanThreshold$Length <- as.vector(seqlengths(diamondGRanges)[pertcentAAGreaterThanThreshold$group_name])
    pertcentAAGreaterThanThreshold <- data.frame(mutate(pertcentAAGreaterThanThreshold,Percentcov=n/Length*100))
    pertcentAAGreaterThanThreshold <- left_join(data.frame(group_name = names(read.fasta(proteinFile))),pertcentAAGreaterThanThreshold)
    pertcentAAGreaterThanThreshold$Percentcov[is.na(pertcentAAGreaterThanThreshold$Percentcov)] = 0
    return(pertcentAAGreaterThanThreshold)
  
}

getModeGranges <- function(diamondGRanges,proteinFile){
  coverage <- as.data.frame(coverage(diamondGRanges))
  totalCoverage <- calculateTotalCoverage(coverage)
  threholdCoverage <- getmode(totalCoverage[,1])
  return(threholdCoverage)
}


load("../app/LangCelegansBuscoGranges.Rdata")
LangeCelegans <- PercentAAGreaterThan(diamondGRanges,"../Similarity Analysis/BuscoCelegans.fasta")
LangeCelegansCV <- getModeGranges(diamondGRanges,"../Similarity Analysis/BuscoCelegans.fasta")

load("../app/LangBuscoGranges.Rdata")
LangBusco <- PercentAAGreaterThan(diamondGRanges,"../Similarity Analysis/BuscoProteinSet.fa")
LangBuscoCV <- getModeGranges(diamondGRanges,"../Similarity Analysis/BuscoProteinSet.fa")

load("../app/SchwarzCelegansBuscoGranges.Rdata")
SchwarzeCelegans <- PercentAAGreaterThan(diamondGRanges,"../Similarity Analysis/BuscoCelegans.fasta")
SchwarzeCelegansCV <- getModeGranges(diamondGRanges,"../Similarity Analysis/BuscoCelegans.fasta")

load("../app/SchwarzBuscoGranges.Rdata")
SchwarzBusco <- PercentAAGreaterThan(diamondGRanges,"../Similarity Analysis/BuscoProteinSet.fa")
SchwarzBuscoCV <- getModeGranges(diamondGRanges,"../Similarity Analysis/BuscoProteinSet.fa")

load("../app/LangCelegansMoreSensativeBuscoGranges.Rdata")
LangeCelegansMoreSensative <- PercentAAGreaterThan(diamondGRanges,"../Similarity Analysis/BuscoCelegans.fasta")
LangeCelegansMoreSensativeCV <- getModeGranges(diamondGRanges,"../Similarity Analysis/BuscoCelegans.fasta")

load("../app/LangCelegansMMseqs2Granges.Rdata")
LangeCelegansMMseqs2 <- PercentAAGreaterThan(diamondGRanges,"../Similarity Analysis/BuscoCelegans.fasta")
LangeCelegansMMseqs2CV <- getModeGranges(diamondGRanges,"../Similarity Analysis/BuscoCelegans.fasta")



ggplot(LangeCelegans,aes(Percentcov))+geom_histogram(binwidth = 5)+ylim(0,110)+labs(title = "Laing sequencing reads aligned to C.elegans BUSCO proteins using DIAMOND", subtitle = paste("CoverageThreshold =",LangeCelegansCV), x = paste("Percent of amino acids with minimum depth >= ",LangeCelegansCV), y = "Count")
ggsave("Lang1.png",device="png", width = 9, height = 7)

ggplot(LangBusco,aes(Percentcov))+geom_histogram(binwidth = 5)+ylim(0,110) +labs(title = "Laing sequencing reads aligned to BUSCO proteins using DIAMOND",subtitle = paste("CoverageThreshold =",LangBuscoCV), x = paste("Percent of amino acids with minimum depth >= ",LangBuscoCV), y = "Count")
ggsave("Lang2.png",device="png", width = 9, height = 7)

ggplot(LangeCelegansMoreSensative,aes(Percentcov))+geom_histogram(binwidth = 5)+ylim(0,110) + labs(title = "Laing sequencing reads aligned to C.elegans BUSCO proteins using DIAMOND More Sensitive",subtitle = paste("CoverageThreshold =",LangeCelegansMoreSensativeCV), x = paste("Percent of amino acids with minimum depth >= ",LangeCelegansMoreSensativeCV), y = "Count")
ggsave("Lang3.png",device="png", width = 9, height = 7)


ggplot(LangeCelegansMMseqs2,aes(Percentcov))+geom_histogram(binwidth = 5)+ylim(0,110) + labs(title = "Laing sequencing reads aligned to C.elegans BUSCO Proteins using MMseqs2",subtitle = paste("CoverageThreshold =",LangeCelegansMMseqs2CV), x = paste("Percent of amino acids with minimum depth >= ",LangeCelegansMMseqs2CV), y = "Count")
ggsave("Lang4.png",device="png", width = 9, height = 7)

mean(LangeCelegans$Percentcov)
mean(LangBusco$Percentcov)
mean(LangeCelegansMoreSensative$Percentcov)
mean(LangeCelegansMMseqs2$Percentcov)


ggplot(LangeCelegans,aes(x=1,y=Percentcov))+geom_violin(alpha=0.2)+coord_flip()+geom_jitter(shape=16,size=0.4) + labs(title = "Laing Aligned  to C.elegans Busco Proteins",subtitle = paste("CoverageThreshold =",LangeCelegansCV))

ggplot(LangBusco,aes(x=1,y=Percentcov))+geom_violin(alpha=0.2)+coord_flip()+geom_jitter(shape=16,size=0.4) + labs(title = "Laing Aligned  to Busco Proteins",subtitle = paste("CoverageThreshold =",LangBuscoCV))

ggplot(LangeCelegansMoreSensative,aes(x=1,y=Percentcov))+geom_violin(alpha=0.2)+coord_flip()+geom_jitter(shape=16,size=0.4)+ labs(title = "Laing Aligned  to C.elegans Busco Proteins Using More Sensitive",subtitle = paste("CoverageThreshold =",LangeCelegansMoreSensativeCV))

ggplot(LangeCelegansMMseqs2,aes(x=1,y=Percentcov))+geom_violin(alpha=0.2)+coord_flip()+geom_jitter(shape=16,size=0.4)+ labs(title = "Laing Aligned to C.elegans Busco Proteins Using MMseqs2",subtitle = paste("CoverageThreshold =",LangeCelegansMMseqs2CV))




ggplot(SchwarzeCelegans,aes(Percentcov))+geom_histogram(binwidth = 5)+ylim(0,80) + labs(title = "Schwarz Aligned to C.elegans Busco Proteins",subtitle = paste("CoverageThreshold =",SchwarzeCelegansCV))

ggplot(SchwarzBusco,aes(Percentcov))+geom_histogram(binwidth = 5)+ylim(0,80) + labs(title = "Schwarz Aligned to Busco Proteins",subtitle = paste("CoverageThreshold =",SchwarzBuscoCV))



mean(SchwarzeCelegans$Percentcov)
mean(SchwarzBusco$Percentcov)


ggplot(SchwarzeCelegans,aes(x=1,y=Percentcov))+geom_violin(alpha=0.2)+coord_flip()+geom_jitter(shape=16,size=0.4) + labs(title = "Schwarz Aligned to C.elegans Busco Proteins",subtitle = paste("CoverageThreshold =",SchwarzeCelegansCV))

ggplot(SchwarzBusco,aes(x=1,y=Percentcov))+geom_violin(alpha=0.2)+coord_flip()+geom_jitter(shape=16,size=0.4) + labs(title = "Schwarz Aligned to Busco Proteins",subtitle = paste("CoverageThreshold =",SchwarzBuscoCV))




```

```{r}

saveFiles <- function(x){
  lapply(split(x,x$file),function(x){
    print(paste0(seqnames(x)[1],".",x$file[1],".txt"))
    write.table(x$Query,paste0("readAllignmentsMMSEQ2/",seqnames(x)[1],".",x$file[1],".txt"),quote=FALSE,row.names = FALSE,col.names = FALSE)
    #,"",sapply(strsplit(x$Query,".",TRUE), "[[", 2),"length=100"
  })
  return("Passed")
}


load("../app/LangCelegansMMseqs2Granges.Rdata")
splitDiamond <- split(diamondGRanges,seqnames(diamondGRanges))
lapply(splitDiamond,saveFiles)

  
```